### ISSUE TITLE
Python 3 support

### ISSUE THREAD

**Issue by oTree-org** [oTree-org/otree-core#296]
I am not very knowledgeable about Python 3 so I don't have an opinion yet on when oTree should support Python 3, and also whether we should at some point drop Python 2 support. 

I am logging this issue to keep track of various considerations.

Pros:
- oTree is more of an isolated framework rather then a component that you plug into or combine with other apps. This isolation gives us more freedom to only support specific versions of Python.
- Even though python 3 is not very popular now, it is eventually supposed to become the standard python version. Now may be a good time to jump on the Python 3 bandwagon.
- It would be beneficial if all oTree apps were written in the same version of Python (Python 3), because that would make it easier to share code, test, provide code samples, etc.
- Python 3 is supposed to be an improvement over Python 2 in terms of language design.

Cons:
- The vast majority of Python code is written in Python 2. This could make it confusing for beginner oTree programmers who are searching for help on Google, stack overflow, etc.
- not sure if all of our dependencies support python 3
- what if Python 3 adoption does not improve much in the next couple of years?

Please feel free to comment or add considerations.

**Comment by leliel12** [comment]
python 3 is better than python 2 and eventually will be the dominant
version. However in our case we need to think in:

- We dont neeed to abandon py2. Django itself use the package [six](
https://pythonhosted.org/six/) to write multiversion code
- Before to decide use six wee need to check if ALL aour dependencies
support python 3

After that i sugest to start a migration

2015-03-29 23:37 GMT-03:00 oTree-org <notifications@github.com>:

> I am not very knowledgeable about Python 3 so I don't have an opinion yet
> on when oTree should support Python 3, and also whether we should at some
> point drop Python 2 support.
>
> I am logging this issue to keep track of various considerations.
>
> Pros:
>
>    - oTree is more of an isolated framework rather then a component that
>    you plug into or combine with other apps. This isolation gives us more
>    freedom to only support specific versions of Python.
>    - Even though python 3 is not very popular now, it is eventually
>    supposed to become the standard python version. Now may be a good time to
>    jump on the Python 3 bandwagon.
>    - It would be beneficial if all oTree apps were written in the same
>    version of Python (Python 3), because that would make it easier to share
>    code, test, provide code samples, etc.
>    - Python 3 is supposed to be an improvement over Python 2 in terms of
>    language design.
>
> Cons:
>
>    - The vast majority of Python code is written in Python 2. This could
>    make it confusing for beginner oTree programmers who are searching for help
>    on Google, stack overflow, etc.
>    - not sure if all of our dependencies support python 3
>    - what if Python 3 adoption does not improve much in the next couple
>    of years?
>
> Please feel free to comment or add considerations.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/oTree-org/otree-core/issues/296>.
>



-- 
Juan B Cabral

**Comment by leliel12** [comment]
Interesting link http://www.randalolson.com/2015/01/30/python-usage-survey-2014/

**Comment by leliel12** [comment]
We **can't** port otree-core to python 3 because: https://caniusepython3.com/project/otree-core

**Comment by oTree-org** [comment]
Thanks for investigating. Some of these packages could easily be made to work with Py3, others I'm not sure, or maybe difficult. Here is the list grouped by a guess of easy it is to get Py3 support, either by porting ourselves or logging a bug with project owners. Maybe some of these projects are already almost Py3 compatible but just not officially. However, we would also need to make sure to check recursively the dependencies of these projects.

## Easy
- django-ptree-extra-views: Unused dependency, should remove
- django-ptree-mturk: Unused dependency, should remove
- ordered-set: Minor (removable) dependency and it's a pretty small library: https://github.com/LuminosoInsight/ordered-set/blob/master/ordered_set.py
- django-easymoney: Was written specifically for oTree, I can ask the author to update. https://github.com/Suor/django-easymoney/blob/master/easymoney.py

## Medium
- django-save-the-change: https://github.com/karanlyons/django-save-the-change/blob/master/save_the_change/mixins.py
- django-sslserver: https://github.com/teddziuba/django-sslserver/blob/master/sslserver/management/commands/runsslserver.py

## Hard
- ipy: Looks kind of complex: https://github.com/haypo/python-ipy/blob/master/IPy.py
- handy: looks like a lot of code: https://github.com/Suor/handy/tree/master/handy

**Comment by leliel12** [comment]
"can i use python 3" check the dependencies recursively

2015-03-30 2:25 GMT-03:00 oTree-org <notifications@github.com>:

> Thanks for investigating. Some of these packages could easily be made to
> work with Py3, others I'm not sure, or maybe difficult. Here is the list
> grouped by a guess of easy it is to get Py3 support, either by porting
> ourselves or logging a bug with project owners. Maybe some of these
> projects are already almost Py3 compatible but just not officially.
> However, we would also need to make sure to check recursively the
> dependencies of these projects.
> Easy
>
>    - django-ptree-extra-views: Unused dependency, should remove
>    - django-ptree-mturk: Unused dependency, should remove
>    - ordered-set: Minor (removable) dependency and it's a pretty small
>    library:
>    https://github.com/LuminosoInsight/ordered-set/blob/master/ordered_set.py
>    - django-easymoney: Was written specifically for oTree, I can ask the
>    author to update.
>    https://github.com/Suor/django-easymoney/blob/master/easymoney.py
>
> Medium
>
>    - django-save-the-change:
>    https://github.com/karanlyons/django-save-the-change/blob/master/save_the_change/mixins.py
>    - django-sslserver:
>    https://github.com/teddziuba/django-sslserver/blob/master/sslserver/management/commands/runsslserver.py
>
> Hard
>
>    - ipy: Looks kind of complex:
>    https://github.com/haypo/python-ipy/blob/master/IPy.py
>    - handy: looks like a lot of code:
>    https://github.com/Suor/handy/tree/master/handy
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/oTree-org/otree-core/issues/296#issuecomment-87552184>
> .
>



-- 
Juan B Cabral

**Comment by gregmuellegger** [comment]
My take on Python 3 support for oTree is, that you absolutely should support Python 3 at some point. The Python ecosystem is slowly but steadily moving from Python 2 to Python 3. However this migration is already years in progress and will still be in progress for the next ~3-5 years. So there is no immediate rush in order to keep up with the Python world.

As a main indication that it really gets necessary you should see what the default Python versions are in the existing Operating System since it's easier to try out oTree if you don't need to install another Python version. I think MacOS and Ubuntu are using Python 2 as a default so far. A few other Linux Distributions already using Python 3 as the default.
@oTree-org knows probably more about the issues on Windows. I would argue that novice Python users would probably install Python 3 as users ofcourse tend to use the latest versions.

In my experience, as long as the dependencies do support Python 3, it's fairly easy to maintain support for Python 2 and Python 3 at the same time using the exact same codebase. A prequisite for this is having a reliable testsuite which is kind of there with oTree (especially considering the integration tests in the oTree library repository).

**Comment by leliel12** [comment]
I other hand the main problem with oTree + py2 is the launcher... and not
only py2 its windows to be more extrcit



2015-03-30 11:44 GMT-03:00 Gregor Müllegger <notifications@github.com>:

> My take on Python 3 support for oTree is, that you absolutely should
> support Python 3 at some point. The Python ecosystem is slowly but steadily
> moving from Python 2 to Python 3. However this migration is already years
> in progress and will still be in progress for the next ~3-5 years. So there
> is no immediate rush in order to keep up with the Python world.
>
> As a main indication that it really gets necessary you should see what the
> default Python versions are in the existing Operating System since it's
> easier to try out oTree if you don't need to install another Python
> version. I think MacOS and Ubuntu are using Python 2 as a default so far. A
> few other Linux Distributions already using Python 3 as the default.
> @oTree-org <https://github.com/oTree-org> knows probably more about the
> issues on Windows. I would argue that novice Python users would probably
> install Python 3 as users ofcourse tend to use the latest versions.
>
> In my experience, as long as the dependencies do support Python 3, it's
> fairly easy to maintain support for Python 2 and Python 3 at the same time
> using the exact same codebase. A prequisite for this is having a reliable
> testsuite which is kind of there with oTree (especially considering the
> integration tests in the oTree library repository).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/oTree-org/otree-core/issues/296#issuecomment-87706853>
> .
>



-- 
Juan B Cabral

**Comment by leliel12** [comment]
IPy works on py3

**Comment by hayd** [comment]
ordered-set also works on py3.
I just did a PR (now merged) for handy.

**Comment by oTree-org** [comment]
Thank you very much @hayd !

**Comment by leliel12** [comment]
i going to take this as personal! i will use six as  all django ecosystem

**Comment by oTree-org** [comment]
OK great :)

**Comment by leliel12** [comment]
I made some researche and create a new py3 branch. Actually only added some changes in setup.py and created a test (like the test_style) for auto check every file in our project if some code are wrong.

**Comment by leliel12** [comment]
Well i hit a bug in django 1.7 itself with metaclasses, first i need to migrate to django 1.8 to proceed with py3k compatibility

**Comment by gregmuellegger** [comment]
I re-initiated the work on Python 3 in the `python3-support` branch.

I already got a far way in, but we are not there yet. There are still some errors when running the test suite.

- One error is related to django-idmap. I filed a bug against their package: https://bitbucket.org/tkhyn/django-idmap/issues/2/python-3-support-in-querysetpy
last time I did this, the maintainer was really quick to fix this.
- I removed the dependency on django-handy and just copied over the JSONField and PickleField into otree itself. I did this as there were some Python 3 incompatibilities in the handy package. I didn't want to bother creating a patch for handy, since there is not really a testsuite for the project anyway. I wouldn't rely on it in otree in the future.
- I worked on Python 3 compatibility for easymoney. I'm mostly there but the locale handling for outputting the correct currency format still bugs me a little. I have created pull request, but it seems it needs some more extra work. PR: https://github.com/Suor/django-easymoney/pull/2

I'll add new comments here in the future when I'm able to make more progress.

**Comment by gregmuellegger** [comment]
The python3-support branch passes all tests for the Python 2 related tests. I would suggest that we merge the python3-support branch directly in master anytime soon to avoid that it will go stale soon. @oTree-org what do you think?

I only see one backwards incompatibility: In order to fix the JSONField for Python 3, I had to change the way it serializes it's content. So this change might make existing database data invalid, so users would need to reset their db when upgrading otree-core.

**Comment by oTree-org** [comment]
Great! I merged it.

**Comment by leliel12** [comment]
Probably you already know this but tox already has several tests for
python3.

tox -e py3.4-XXXXX

and
tox -e 2to3

On Tue, 12 Jan 2016 at 15:02 oTree-org <notifications@github.com> wrote:

> Great! I merged it.
>
> Sent from my iPhone
>
> > On Jan 13, 2016, at 2:41 AM, Gregor Müllegger <notifications@github.com>
> wrote:
> >
> > The python3-support branch passes all tests for the Python 2 related
> tests. I would suggest that we merge the python3-support branch directly in
> master anytime soon to avoid that it will go stale soon. @oTree-org what do
> you think?
> >
> > I only see one backwards incompatibility: In order to fix the JSONField
> for Python 3, I had to change the way it serializes it's content. So this
> change might make existing database data invalid, so users would need to
> reset their db when upgrading otree-core.
> >
> > —
> > Reply to this email directly or view it on GitHub.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/oTree-org/otree-core/issues/296#issuecomment-170996060>
> .
>
-- 

Juan BC (from phone)

**Comment by gregmuellegger** [comment]
Yep cool. Cheers for the ground works there!

**Comment by gregmuellegger** [comment]
I worked more on the Python 3 compatibility. I was able to perform the following things:

- The unittests now pass with Python 3, using the easymoney and idmap updates mentioned below.
- I resolved all warnings found by `py3kwarn`.
- I found an issue in the combination of django-easymoney and Babel on Python 3. I resolved it in a new PR for django-easymoney.
- I nudged the django-idmap creator for a new release so we can use the Python 3 fix I provided.

Things that are currently on my todo, but I can only continue with those once the blockers from below are resolved:

- Test that django-sslserver works with Python 3 (caniusepython3 reports it as Python 2 only so far)
- Test otree-save-the-change properly on Python 3 (caniusepython3 reports it as Python 2 only so far)

I see the following things as blockers before we can call otree-core to be Python 3 compatible:

- django-easymoney needs to merge this PR and issue a new release: https://github.com/Suor/django-easymoney/pull/3
- Once there is a new release of django-easymoney, we need to use the version 2.2.0 of Babel. I have created a branch that works with Babel 2.2.0: https://github.com/oTree-org/otree-core/tree/update-babel
- django-idmap needs to issue a new release. See https://bitbucket.org/tkhyn/django-idmap/issues/2/python-3-support-in-querysetpy (not in our control, but I assume it will happen soonish as the maintainer is very quick in responding)

**Comment by gregmuellegger** [comment]
@oTree-org Could you nudge @Suor for look at the PR and release in django-easymoney at some point? Currently that's a blocker for oTree's Python 3 support.

**Comment by oTree-org** [comment]
ok asked him

**Comment by PhoenixAlx** [comment]
How is the issue? @gregmuellegger 

> 
> I worked more on the Python 3 compatibility. I was able to perform the following things:
> 
>     The unittests now pass with Python 3, using the easymoney and idmap updates mentioned below.
>     I resolved all warnings found by py3kwarn.
>     I found an issue in the combination of django-easymoney and Babel on Python 3. I resolved it in a new PR for django-easymoney.
>     I nudged the django-idmap creator for a new release so we can use the Python 3 fix I provided.
> 
> Things that are currently on my todo, but I can only continue with those once the blockers from below are resolved:
> 
>     Test that django-sslserver works with Python 3 (caniusepython3 reports it as Python 2 only so far)
>     Test otree-save-the-change properly on Python 3 (caniusepython3 reports it as Python 2 only so far)
> 
> I see the following things as blockers before we can call otree-core to be Python 3 compatible:
> 
>     django-easymoney needs to merge this PR and issue a new release: Suor/django-easymoney#3
>     Once there is a new release of django-easymoney, we need to use the version 2.2.0 of Babel. I have created a branch that works with Babel 2.2.0: https://github.com/oTree-org/otree-core/tree/update-babel
>     django-idmap needs to issue a new release. See https://bitbucket.org/tkhyn/django-idmap/issues/2/python-3-support-in-querysetpy (not in our control, but I assume it will happen soonish as the maintainer is very quick in responding)
> 
>

**Comment by gregmuellegger** [comment]
I had two made two small changes in oTree library to integrate Python 3 support: oTree-org/oTree@4f988e0 and oTree-org/oTree@965ca8d. I have already pushed them to master. Otherwise the library seems to work perfectly well with Python 3 in its current state.

All other changes were done in otree-core. I upgraded django-easymoney, babel and django-idmap. I activated the TravisCI tests for Python 3.5 as well so that we make sure not to break Python 3 support in future versions.

There were two packages that are not marked as Python 3 compatible on PyPI, but my tests told me that they are perfectly fine to use:

- django-sslserver: my tests verify that it works on Python 3, there is also an already merged PR for the project that states it supports Python 3 https://github.com/teddziuba/django-sslserver/pull/23
- otree-save-the-change: I have not done any specific testing about this, but since all tests are passing, I assume it works the way we want it to work.

**I consider otree-core master and oTree library master now as Python 3 compatible**. I would suggest as next step that @oTree-org does some manual testing and see if everything behaves as expected.

Tests for Python 3 are in place and passing: https://travis-ci.org/oTree-org/otree-core/builds/110985017
